### PR TITLE
Directus extension SDK type fix

### LINKED ISSUE #23832: Typescript errors after upgrading
### Describe the Bug

Ever since upgrading Directus to 11.1.1 we are experiencing typescript issues. Seems like the hook definition types have been changed?

These are the errors we experiencing with a newly created extension:
Screenshot 2024-10-10 at 10 30 22
Screenshot 2024-10-10 at 10 34 02

### To Reproduce

Create a new extension with `npx create-directus-extension@latest` and try to compile. Should give the exact same errors.

### Directus Version

11.1.1

### Hosting Strategy

Self-Hosted (Docker Image)

### Database

pg16

### PR DESCRIPTION
**Pull request by br41nslug**
Fixes #23832 

## Scope

What's changed:

This reverts specifically the backend extension type changes made in https://github.com/directus/directus/pull/23714

The `ExtendedConfig` generic type does not work for config definitions which can be functions as both endpoints and hooks allow.

I've also added `T extends object` in the `CustomConfig` and `ExtendedConfig` generics because they both rely on `keyof T` which is only supported on `object` types.

## Potential Risks / Drawbacks

- None that im aware of, types that dont use the `ExtendedConfig` logic have been reverted and simplified

## Review Notes / Questions

- I've kept this PR intentionally small, only fixing the backend `defineX` functions which failed as they can be functions instead of objects

### CODE REVIEW

File: packages/extensions/src/shared/utils/define-extension.ts [hunk 1]
```diff
@@ -11,18 +11,18 @@ import type {
 	PanelConfig,
 } from '../types/index.js';
 
-type CustomConfig<T> = { [K in string]: K extends keyof T ? never : unknown };
+type CustomConfig<T extends object> = Record<string, unknown> & T;
```
Comment (br41nslug) [inline packages/extensions/src/shared/utils/define-extension.ts:14]: Reminder to self: discuss experimental type changes with hannes

**Review by paescuj — APPROVED** [review]
LGTM 🚀

**Comment by keesvanbemmel** [comment]
Thanks for this @br41nslug  @paescuj ! Any guestimates on when Next Release will be?

**Comment by br41nslug** [comment]
No promises or guarantees but we generally try to do a 2 week release cycle

**Pull request merged**